contracts/smart-account/modules/SessionValidationModules/ABISessionValidationModule.sol

```diff
@@ -0,0 +1,239 @@
+                bytes4(callData[0:4]) == EXECUTE_SELECTOR,
+            "ABISV Not Execute Selector"
+        );
+
+        address destContract;
+        uint256 callValue;
+        bytes calldata data;
+        assembly {
+            destContract := calldataload(add(callData.offset, 0x4))
+            callValue := calldataload(add(callData.offset, 0x24))
+
+            let dataOffset := add(
+                add(callData.offset, 0x04),
+                //offset of the bytes arg is stored after selector and two first 32-byte args
+                calldataload(add(callData.offset, 0x44))
+            )
+
+            let length := calldataload(dataOffset)
+            data.offset := add(dataOffset, 0x20)
+            data.length := length
+        }
+
+        return
+            _validateSessionParams(
+                destContract,
+                callValue,
+                data,
+                _sessionKeyData
+            ) ==
+            ECDSA.recover(
+                ECDSA.toEthSignedMessageHash(_userOpHash),
+                _sessionKeySignature
+            );
+    }
+
+    /**
+     * @dev validates that the call (destinationContract, callValue, funcCallData)
+     * complies with the Session Key permissions represented by sessionKeyData
+     * @param destinationContract address of the contract to be called
+     * @param callValue value to be sent with the call
+     * @param _funcCallData the data for the call. is parsed inside the SVM
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * param _callSpecificData additional data, specific to the call, not used here
+     * @return sessionKey address of the sessionKey that signed the userOp
+     * for example to store a list of allowed tokens or receivers
+     */
+    function validateSessionParams(
+        address destinationContract,
+        uint256 callValue,
+        bytes calldata _funcCallData,
+        bytes calldata _sessionKeyData,
+        bytes memory /*_callSpecificData*/
+    ) external pure virtual override returns (address) {
+        return
+            _validateSessionParams(
+                destinationContract,
+                callValue,
+                _funcCallData,
+                _sessionKeyData
+            );
+    }
+
+    /**
+     * @dev validates that the call (destinationContract, callValue, funcCallData)
+     * complies with the Session Key permissions represented by sessionKeyData
+     * @param destinationContract address of the contract to be called
+     * @param callValue value to be sent with the call
+     * @param _funcCallData the data for the call. is parsed inside the SVM
+     * @param _sessionKeyData SessionKey data, that describes sessionKey permissions
+     * @return sessionKey address of the sessionKey that signed the userOp
+     * for example to store a list of allowed tokens or receivers
+     */
+    function _validateSessionParams(
```

Comment:
can you add more comments on solidity assembly blocks about what it achieves.
also would be great is constant numbers can be defined as constants or some offset constant then x:x+4 (depending on bytes)

Reply:
added

Reply:
added couple of constants, in other places just added comments not have constant+constant (which would add gas) instead of just hardcoded value